A1.md

```diff
@@ -0,0 +1,133 @@
+It accepts a TCP connection from a client, creates a new TCP connection
+to another server, and forwards all bytes from the client to the server,
+without actually understanding anything about the contents of those bytes.
+
+Note that load-balancing within gRPC happens on a per-call basis, not
+a per-connection basis.  Because of this, use of a TCP-level proxy may
+limit the ability to do load balancing in some environments.
+
+### Related Proposals: 
+
+N/A
+
+## Proposal
+
+gRPC will support TCP-level proxies via the HTTP CONNECT request,
+defined in [RFC-2817](https://tools.ietf.org/html/rfc2817).
+
+We are aware of the following use-cases for TCP-level proxying with gRPC:
+
+1. A corp environment where all traffic outbound to the internet must go
+   through a proxy.  This environment has the following properties:
+   - DNS resolution of external names does not work (i.e., all
+     resolution must be done via the proxy).
+   - *All* requests must go through the proxy, both for internal and
+     external servers.
+   - This environment is generally triggered by some centralized signal,
+     such as the `http_proxy` environment variable or the Java registry.
+
+2. A partially protected environment, where access to certain addresses
+   must go through a proxy.  This environment has the following
+   properties:
+   - DNS resolution of protected servers works normally.
+   - Only requests for certain hosts must go through the proxy.
+     Requests to other servers work without the proxy.
+   - Custom logic is used to determine which hosts the proxy will be
+     used for.
+
+We propose the following functionality to support these two cases:
+
+- For case 1, the client channel code will check for the existence of the
+  `http_proxy` environment variable (or equivalent signal).  If that
+  signal is present, the client channel code will do the following:
+  - Ask the resolver to resolve the proxy name instead of the server name
+    requested via the client API.  (Note that this will cause us to
+    create a single subchannel to the proxy, as opposed to one
+    subchannel for each backend server.)
+  - Add state (e.g., a channel argument in C-core) requesting that the
+    subchannel code (see below) send an HTTP CONNECT request for the
+    server name requested via the client API.
+
+- For case 2, in the subchannel, before we connect to the target address,
+  we will call a new *proxy mapper* hook, which will allow selectively
+  requesting the use of a proxy based on the address to which the
+  subchannel is going to connect.  If a proxy is to be used, the mapper
+  will return the proxy address (which the subchannel will connect to
+  instead of the address returned by the resolver) and some additional
+  state (e.g., a channel argument in C-core) requesting that the
+  subchannel send an HTTP CONNECT request for the original subchannel
+  address.  Note that in this case, the server name used in the HTTP
```

Comment:
The proxy mapper could decide what type of HTTP CONNECT request is going to be made: it doesn't have to be to the original address: it could be to the original name instead.